Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs

```diff
@@ -136,274 +194,253 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
-#endif
+                    while (!disposingCancellationTokenSource.IsCancellationRequested)
+                    {
+                        await WaitForPendingRequest();
 
-                    throw new OperationCanceledException();
+                        await ProcessPendingRequest();
+                    }
                 }
-
-                Page candidatePage = await TryLoadNextOrWaitForPreLoadingToFinish(precedingCheckpoint, subscriptionId)
-                    .ConfigureAwait(false);
-
-                if (candidatePage.PrecedingCheckpoint == precedingCheckpoint)
+                catch (OperationCanceledException)
                 {
-                    return candidatePage;
+                    // The cancellation token was triggered, so let this path die off silently.                  
                 }
-            }
+            }, disposingCancellationTokenSource.Token);
         }
 
-        private Task<Page> TryLoadNextOrWaitForPreLoadingToFinish(long precedingCheckpoint, string subscriptionId)
+        private async Task WaitForPendingRequest()
         {
-            if (isDisposed)
-            {
-                return Task.FromResult(new Page(precedingCheckpoint, new Transaction[0]));
-            }
-
-            TaskCompletionSource<Page> ourPreloader = null;
-            bool isOurPreloader = false;
-            Task<Page> loader = Volatile.Read(ref currentPreloader);
-
-            try
+            if (pendingRequests.IsEmpty)
             {
-                if (loader == null)
+                await requestQueued.Task;
+                if (pendingRequests.IsEmpty)
                 {
-                    ourPreloader = new TaskCompletionSource<Page>();
-                    Task<Page> oldLoader = Interlocked.CompareExchange(ref currentPreloader, ourPreloader.Task, null);
-                    isOurPreloader = (oldLoader == null);
-                    loader = isOurPreloader ? ourPreloader.Task : oldLoader;
-                }
+                    requestQueued = new TaskCompletionSource<bool>();
 
-                return loader;
+                    await requestQueued.Task.WithWaitCancellation(disposingCancellationTokenSource.Token);
+                }
             }
-            finally
+        }
+
+        private async Task ProcessPendingRequest()
+        {
+            // Keep the request on the queue, so successive subscribers for the same page can "await"
+            // the result of the existing request
+            if (pendingRequests.TryPeek(out LoadRequest request))
             {
-                if (isOurPreloader)
+                try
                 {
-#if LIQUIDPROJECTIONS_DIAGNOSTICS
-                    logger(() => $"Subscription {subscriptionId} created a loader {loader.Id} " +
-                                     $"for a page after checkpoint {precedingCheckpoint}.");
-#endif
-
-                    if (isDisposed)
+                    if (IsMajorityCached(request.PrecedingCheckpoint))

```

Comment:
Additionally, if the first 40% is in the cache, you only need to request that part …
I understand this would not be easy to implement, and might not be worth it

Reply:
Nice, but not now.